File: base_demo/scripts/service_server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import rclpy
from rclpy.node import Node
from base_demo.srv import SetTargetDetec

class MinimalService(Node):

    def __init__(self):
        super().__init__('service_server')
        self.get_logger().info('service_server python node is Ready!')
        self.srv = self.create_service(SetTargetDetec, 'target_detection', self.target_detection_handle)

    def target_detection_handle(self, request, response):
        # 打印输出req.name
        self.get_logger().info('Target object is ' + request.name)
        self.get_logger().info('Find ' + request.name + ' and response.')
        # 对res的成员进行赋值
        response.success = True
        response.pose.position.x = 0.35
        response.pose.position.y = -0.35
        response.pose.position.z = 0.1
        response.pose.orientation.w = 1.0
        return response


def main():
    rclpy.init()
    minimal_service = MinimalService()
    rclpy.spin(minimal_service)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: base_demo/src/service_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "base_demo/srv/set_target_detec.hpp"

// 服务回调函数,req为服务的请求,res为服务的应答
void targetDetectionHandle(const std::shared_ptr<base_demo::srv::SetTargetDetec::Request> req,
          std::shared_ptr<base_demo::srv::SetTargetDetec::Response>   res){
  // 打印输出req.name
  RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"),"Target object is " << req->name);
  RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"),"Find " << req->name << " and response.");
  // 对res的成员进行赋值并返回应答
  res->success = true;
  res->pose.position.x = 0.35;
  res->pose.position.y = -0.35;
  res->pose.position.z = 0.1;
  res->pose.orientation.w = 1;
}

int main(int argc, char **argv){
  // 初始化ROS节点
  rclcpp::init(argc, argv);
  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("service_server");

  rclcpp::Service<base_demo::srv::SetTargetDetec>::SharedPtr service =
    node->create_service<base_demo::srv::SetTargetDetec>("target_detection", &targetDetectionHandle);

  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Service server is Ready!");

  rclcpp::spin(node);
  rclcpp::shutdown();
}
